module system: Improve error messages around faulty imports

roberth opened the pull request:
###### Description of changes

- Make explicit the module "class", which identifies the module system application, such as `"nixos"`, `"darwin"` or `"homeManager"`. This allows us to print an appropriate error when importing for example a home-manager module into NixOS:

  `The module foo.nix#darwinModules.default was imported into nixos instead of home-manager.`

  I do not expect users or module authors to set the `class` attribute in their modules (although they may). Instead I envision a library like [flake-parts](https://flake.parts) to set this for us.

- ~Use the new class information to automatically import the appropriate default module from a flake.~

- Detect other mistakes that involve the importing `_type`-carrying values. For instance importing a `nixosSystem` result into NixOps network appears to be a somewhat common mistake, and for good reason. These changes detect this scenario and explain what to do instead.

Depends on https://github.com/NixOS/nix/issues/7186

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth committed:
lib/modules: Add class concept to check imports

This improves the error message when an incompatible module is
imported.

shlevy reviewed (dismissed):
NACK from me. This is non-compositional and bloats the evalModules arguments, there is a much simpler solution https://github.com/NixOS/nixpkgs/issues/196584#issuecomment-1282602545

roberth commented:
> This is non-compositional

I'm not sure I follow. You may be referring to either commit.
 - If you refer to the introduction and checking of `class`: The purpose of nominal typing is to prevent bad compositions without having to both the user with irrelevant details.
 - If you refer to the treatment of flake attrsets in the module system: I've already proposed an alternative that did not couple with flakes, but that was shot down because allowing arbitrary manipulation of all modules would allow the possibility of behaviors that make modules harder to understand.

> bloats the evalModules arguments

I disagree. It is optional and it will only be set in contexts where `imports` are expected to be used. For example, NixOS sets it only once per system, counting specialisations as systems of course. Submodules that provide the various `foo.<name>.{x,y,z}` don't need to set it.

`class` does have some use as a special arg, as it allows conditionals in modules that are otherwise identical between, say, `nixos` and `darwin`.

> there is a much simpler solution [#196584 (comment)](https://github.com/NixOS/nixpkgs/issues/196584#issuecomment-1282602545)

That's unnecessary logic. I'd rather read some extra attribute paths than some custom `map` call that doesn't transfer between classes.

shlevy commented:
I was blurring things together, let me separate them.

Re `class`, I think this is a good idea, but it doesn't really make sense to pass it in `specialArgs`. It's not (just) an argument passed to each module, it's a change to how modules are interpreted, so it should be a top-level argument to `evalModules` (with `null` default), and if it is set it can be passed to each module as well if desired.

My real objection, and the lack of modularity, is in hard-coding the `m.${classModules}.default` lookup. This is setting opinionated policy rather than providing composable mechanism, ironically enough a policy that you just [disagreed with](https://github.com/hercules-ci/flake-parts/pull/61#issuecomment-1294954964) in one particular instance, and becomes a piece of magic that cannot be overridden nor inferred from the usage site. I still think the right way to handle this is with a `map` in the end-user's invocation, but it certainly doesn't need to be in a core agnostic component like the module system; e.g. we could have the `map` I suggested inside `flake-parts.lib.evalFlakeModule`

roberth commented:
> Re `class`, I think this is a good idea

Cool. I don't really like the name I came up with. I've also considered `kind`, but I feel that's too tied to `value : (type : kind)`, although we don't have that here. `sort` seemed ugly but I'd be easily convinced. Maybe there's a better word?

> but it doesn't really make sense to pass it in `specialArgs`.

It is conveniently low overhead when it comes to most evalModules calls (submodules), but I have not made a measurement.

> this is setting opinionated policy

That is exactly the intent. Some things shouldn't be left to opinion.
This seems to be where we disagree.

> rather than providing composable mechanism

We already have this, and a `map` based solution can still be used as desired. As such it does not enforce a policy, but merely encourages it.

----------

Final question

If we have
 - `nix run` = `nix run apps.default`
 - `nix devshell` = `nix devshell devShells.default`

then wouldn't it make sense to have
 - `imports = [ flake ];` = `imports = [ flake.nixosModules.default ];`

If you agree, and the module system can't know about flake attributes, we'd need a generic mechanism. I this case the disagreement would be with infinisil.
If you disagree, that choice is a small loss for users and a small win for the ivory tower. We'd have to agree to disagree, and I'll leave this pr open, so that others can form an opinion.

shlevy commented:
> It is conveniently low overhead when it comes to most evalModules calls (submodules), but I have not made a measurement.

What overhead do you mean? Syntactically, `{ specialArgs.class = ... }` seems basically equivalent overhead to `{ class = ... }`. In terms of runtime, either we allocate an extra attr on the top-level arg to `evalModules` or we evaluate an extra attr on `specialArgs`. What's the difference?

> That is exactly the intent. Some things shouldn't be left to opinion.

We shouldn't layer flake-specific knowledge into the module system.

> that choice is a small loss for users and a small win for the ivory tower

I understand your disagreement with my stance, but I object to this characterization. The abstract principles I'm applying here, of composable interfaces without layering violations, are principles I hold *because* they result in better software and I'm applying them *because* I think this case is within the context that those principles are sound. The "ivory tower" comment suggests that there is some necessary conflict between theory and practice and I'm choosing theory here, but there is no such conflict.

In this particular case, I only think this is a "loss for users" in a short term sense. The ultimate cost of this decision, and more importantly making decisions like this down the road, is that the code becomes less maintainable, more tightly coupled, more brittle, less usable in new contexts. Especially since the exact same functionality can be moved to components that are actually flake-specific, I see no reason to add this here that wouldn't also result in more and more cruft over time.

roberth commented:
> > It is conveniently low overhead when it comes to most evalModules calls (submodules), but I have not made a measurement.
> 
> What overhead do you mean? Syntactically, `{ specialArgs.class = ... }` seems basically equivalent overhead to `{ class = ... }`. In terms of runtime, either we allocate an extra attr on the top-level arg to `evalModules` or we evaluate an extra attr on `specialArgs`. What's the difference?

`evalModules` would have to handle the argument on every invocation and create an entry for it in an Env, not just when it is set, but in each submodule.

> > That is exactly the intent. Some things shouldn't be left to opinion.
> 
> We shouldn't layer flake-specific knowledge into the module system.

... and `lib` shouldn't have knowledge of systems, derivations, packages shouldn't know about their NixOS tests.
Yet, these violations of the implicit architectural guidelines are very helpful every day. Are those bothering you?

> > that choice is a small loss for users and a small win for the ivory tower
> 
> I understand your disagreement with my stance, but I object to this characterization. The abstract principles I'm applying here, of composable interfaces without layering violations, are principles I hold _because_ I think they result in better software and I'm applying them _because_ I think this case is within the context that those principles are sound. The "ivory tower" comment suggests that there is some necessary conflict between theory and practice and I'm choosing theory here, but there is no such conflict.

We disagree.

> In this particular case, I only think this is a "loss for users" in a short term sense. The ultimate cost of this decision, and more importantly making decisions like this down the road,

Each decision should be judged by its merit.

> is that the code becomes less maintainable, more tightly coupled, more brittle, less usable in new contexts. Especially since the exact same functionality can be moved to components that are actually flake-specific

It can not. It's supposed to be a convenience, not an eyesore. We've already discussed this.

>, I see no reason to add this here that wouldn't also result in more and more cruft over time.

I find this disrespectful. Infinisil and I have been maintaining and improving the module system for the last couple of years, with diligence and care. Why wouldn't we continue to do so?

shlevy commented:
> evalModules would have to handle the argument on every invocation and create an entry for it in an Env, not just when it is set, but in each submodule.

Ah, right. Would be nice to see measurements here but makes sense.

> ... and lib shouldn't have knowledge of systems, derivations, packages shouldn't know about their NixOS tests.
Yet, these violations of the implicit architectural guidelines are very helpful every day. Are those bothering you?

Nix is a language for package management. Why shouldn't lib have knowledge of systems and derivations?

As for the package issue, this does actually bother me a bit since it encourages testing only on NixOS and discourages building a testing interface that is decoupled from the distro. But there is also an important difference between a layering violation in leaf applications of a system and layering violations in a core library component, one which is actually used in many different contexts.

> Each decision should be judged by its merit.

I agree, my point is that a) on this particular case, this is a small step of additional tight coupling and magic surprise and b) the reasoning applied in this case, if applied in future decisions, would equally justify further tight coupling and magic surprise.

> I find this disrespectful. Infinisil and I have been maintaining and improving the module system for the last couple of years, with diligence and care. Why wouldn't we continue to do so?

I wasn't making a comment on your diligence or care, and I appreciate the ongoing work on the module system. I was making a claim about the logical implication of the reasoning being applied here, not about the quality of your efforts. If you disagree, OK, you disagree, I think that you're wrong and that the overall quality of the work will suffer as a result but that doesn't mean I think you're not overall doing an important job well.

roberth committed:
lib/tests/modules.sh: Unload implicit modules

I had some trouble understanding this. Let's try to keep new tests
a bit more stateless and explicit.

roberth committed:
lib/modules: Check against importing things with a _type

roberth committed:
lib/modules: Improve error when a configuration is imported

This is appears to be a fairly common mistake for beginners who want
to build larger things from the system configurations, such as NixOps
networks, etc. Further explanation seems appropriate.

roberth committed:
lib/modules: Explain that a configuration can't be loaded as a module

roberth commented:
I've pivoted the PR to type checking instead of type matching, and it now covers an extra set of errors where we can be quite helpful with our error messages.
No more layer violation.

infinisil commented on lib/modules.nix:
```diff
@@ -397,7 +426,7 @@ rec {
         in parentFile: parentKey: initialModules: args: collectResults (imap1 (n: x:
           let
-            module = loadModule args parentFile "${parentKey}:anon-${toString n}" x;
+            module = checkModule (loadModule args parentFile "${parentKey}:anon-${toString n}" x);
```
I think it's cleaner to use `let in` bindings in e.g. `loadModule` instead of wrapping with another function call here. This way all the processing is together and not split apart. I think functions should only be used when they have more than one use.

infinisil replied:
> the raw module is still in scope

Good point. Maybe something like this would be nicer though:

```nix
loadModule = lib.flip lib.pipe [
  (m: ...)
  (m: ...)
]
```

Anyways, I think it's fine as-is for now

infinisil commented on lib/modules.nix:
```diff
@@ -342,28 +347,52 @@ rec {
+        else if isAttrs m then
+          if m._type or "module" == "module" then
+            unifyModuleSyntax fallbackFile fallbackKey (applyModuleArgsIfFunction fallbackKey m args)
+          else if m._type == "if" || m._type == "override" then
+            loadModule args fallbackFile fallbackKey { config = m; }
```
This is new functionality, I'm not sure I like this idea. For readers of this comment, this is the associated test case for a module that now works:
```nix
{
  imports = [ (lib.mkIf true { enable = true; }) ];
}
```

This adds another type of module that's valid, in addition to files, attribute sets with specific keys, attribute sets with arbitrary keys (implicit `config`), functions returning such attribute sets.

It's also a bit weird in that `mkIf` and `mkForce` (and co.) are only available under `lib`, which wouldn't even be in scope in e.g. a standalone file containing such a module. So this module type is strictly only useful when you already have a `lib` in scope (this also irks me a bit because `lib` contains part of the module system API, and it potentially leads [to problems](https://github.com/NixOS/nixpkgs/pull/98952) if you mix them).

roberth replied:
This isn't new functionality; the test I added already worked before this change. Double checked :)

> irks

We should consider `mkIf` and friends to be a stable interface by now. They are little more than constructors, and I see no reason to change them in the future, unless we want to replace them by something that isn't the module system anymore. For what it's worth; this is basically off-topic considering there's no change in behavior here.

infinisil replied:
> This isn't new functionality; the test I added already worked before this change. Double checked :)

Ahh sorry yes! Yeah nevermind :)

infinisil commented on lib/modules.nix:
```diff
@@ -342,28 +347,52 @@ rec {
       loadModule = args: fallbackFile: fallbackKey: m:
-        if isFunction m || isAttrs m then
+        if isFunction m then
           unifyModuleSyntax fallbackFile fallbackKey (applyModuleArgsIfFunction fallbackKey m args)
```
If there are no other uses of `applyModuleArgsIfFunction`, we should probably deprecate it and define it locally here instead. Because with the `isFunction` check here, the `IfFunction` functionality is always triggered. While for the below `isAttrs` branch, this function never does anything and can be omitted.

infinisil commented on lib/modules.nix:
```diff
@@ -342,28 +347,52 @@ rec {
+          else
+            throw (
+              "Could not load a value as a module, because it is of type ${lib.strings.escapeNixString m._type}"
+              + lib.optionalString (fallbackFile != null) ", in file ${toString fallbackFile}."
```
I think `fallbackFile` is always non-`null`, no?

roberth replied:
That's right. It can be `unknownModule` instead; improved it.

infinisil commented on lib/modules.nix:
```diff
@@ -342,28 +347,52 @@ rec {
         options = checked options;
         config = checked (removeAttrs config [ "_module" ]);
         _module = checked (config._module);
         inherit extendModules type;
       };
     in result;
 
-  # collectModules :: (modulesPath: String) -> (modules: [ Module ]) -> (args: Attrs) -> [ Module ]
+  # collectModules :: (class: String) -> (modulesPath: String) -> (modules: [ Module ]) -> (args: Attrs) -> [ Module ]
   #
   # Collects all modules recursively through `import` statements, filtering out
   # all modules in disabledModules.
-  collectModules = let
+  collectModules = class: let
```
This is technically a breaking change since `collectModule` is exposed under `lib.modules.collectModule`, but that's arguably a design mistake, it should never have been exposed, I'd be in favor of making it or marking it internal somehow.

roberth replied:
The time has come to deprecate the functions we consider to be probably internal. By deprecating them, we can figure out whether anyone uses them, while hopefully relieving us of this supposedly needless contract soon-ish.

infinisil commented on lib/modules.nix:
```diff
@@ -105,6 +105,10 @@ rec {
                   # when resolving module structure (like in imports). For everything else,
                   # there's _module.args. If specialArgs.modulesPath is defined it will be
                   # used as the base path for disabledModules.
+                  #
+                  # `specialArgs.class`:
+                  # A nominal type for modules. When set and non-null, this adds a check to
+                  # make sure that only compatible modules are imported.
                   specialArgs ? {}
```
While I generally like the module class concept, I think it should be thought about a bit more. Here's some questions:
1. What about submodules? Should they be able to use classes too? If yes, what should the classes be called?
2. If we have lots of classes, would we have conflicts? Would we need a convention to avoid them or resolve them? Especially in light of third-party code interacting with each other, not just within Nixpkgs
3. What do these classes exactly do? Can we formally write down the semantics? Currently they don't have any changed behavior other than throw an error in some cases so this might not be that hard.
4. Related: Would we want versioning for this, now or in the future? This is made for third-parties where the Nixpkgs version that gets used can vary greatly, would we need to attach a NixOS release number to the class? Would we need a way for third-parties to provide different modules depending on the NixOS release and the class itself?  
5. Could there be false positives, where users _want_ to have a module shared between two differing classes? I guess this is probably fine since modules can always omit class for the check not to happen

Maybe I'm overly-anxious here, just trying to think of the possibilities

roberth replied:
> 1. What about submodules? Should they be able to use classes too?

This should work the same for submodules.

> If yes, what should the classes be called?

There's basically no different. `class` is still `class` and the values are for the module author to pick. If you use NixOS as a submodule, it's already set for you. If you use a submodule as a record type with no behavior or reuse, you don't need a `class`. If you use a submodule as a place to import modules, you can pick your own name for it. This would be rare in NixOS, but flake-parts would have e.g. `"flake-parts"` and `"flake-parts-perSystem"`, but for example `haskell-flake` has its own class of modules to be imported into haskell project submodules.

> 2\. If we have lots of classes, would we have conflicts? Would we need a convention to avoid them or resolve them? Especially in light of third-party code interacting with each other, not just within Nixpkgs

I've considered this, and decided against acting on this, as it is quite unlikely to cause real issues. Let's keep it simple.




> 3\. What do these classes exactly do? Can we formally write down the semantics?

They restrict which modules can be imported. A module can be imported when its class is `null` or equal to the `class` specified by through `specialArgs`.



> 4\. Related: Would we want versioning for this, now or in the future?

Currently we don't. In the future, we could, if the need arises.
Modules already have a lot of options to adapt to the `specialArgs` and `options` environment they're in, including the possibility to error out. I don't think this functionality needs to be coupled to the `class` feature.



> 5\. Could there be false positives, where users _want_ to have a module shared between two differing classes? I guess this is probably fine since modules can always omit class for the check not to happen

That's also my reasoning. Omission of a class is a feature, not a bug.



> Maybe I'm overly-anxious

I hope I've convinced you not to be :)

roberth committed:
lib/modules.nix: Use explicit exports

roberth committed:
lib/modules.nix: Make some functions private

The supposedly public nature of these functions has been holding
back module system maintenance, while usages of these functions
are expected to be rare. If used anywhere, presumably they're
emulating module system behavior because some use case isn't supported
properly. We should try to support such a use case directly, if it
even exists.

roberth committed:
lib/modules.nix: Restore old collectModules interface

roberth committed:
lib/modules.nix: Refactor: evaluate applyModuleArgsIfFunction in attrs case

roberth committed:
lib/modules.nix: Refactor: extract applyModuleArgs

infinisil commented on lib/modules.nix:
```diff
@@ -461,9 +487,13 @@ rec {
         imports = m.require or [] ++ m.imports or [];
         options = {};
         config = addFreeformType (removeAttrs m ["_file" "key" "disabledModules" "require" "imports" "freeformType"]);
```
```suggestion
        config = addFreeformType (removeAttrs m ["_file" "key" "disabledModules" "require" "imports" "freeformType" "class"]);
```

Otherwise you can't do `imports = [ { class = "nixos" } ]`

roberth replied:
If removed `class` as a keyword from the shorthand syntax. I think we want to keep that syntax for "data-like" usage and not "module-like" usage. Stealing `class` might have caused some breakage.

infinisil commented on lib/modules.nix:
```diff
@@ -256,6 +256,7 @@ rec {
 
       merged =
         let collected = collectModules
+          specialArgs.class or null
```
```suggestion
          (specialArgs.class or null)
```

Technically not necessary but this feels illegal!

infinisil commented on lib/modules.nix:
```diff
@@ -342,28 +343,52 @@ rec {
       };
 
       result = withWarnings {
+        _type = "configuration";
```
```suggestion
        _type = if specialArgs ? class then "${specialArgs.class}-evaluation" else "module-evaluation";
```

Or something along those lines, to be able to distinguish between the different evaluation results, maybe don't even set a type if there's no class.

roberth replied:
This makes matching the type of all configurations harder. For example option docs generation should be able to match all of them.
In the status quo this is easy, as is writing a type checking function that checks both the type and the class.
I've picked "configuration" instead of "evaluation" because it matches the flake attribute conventions.

infinisil commented on lib/modules.nix:
```diff
@@ -451,6 +476,7 @@ rec {
           imports = m.imports or [];
           options = m.options or {};
           config = addFreeformType (addMeta (m.config or {}));
+          class = m.class or null;
```
```suggestion
          class = m._class or null;
```

This feels like something that end users shouldn't set, so maybe it should be indicated as internal.

roberth replied:
I've changed it to a non-shorthand-only attribute, just like `meta`. I don't think it needs to be internal, even if it's possibly rare for users to interact with it. Marking this with an underscore would lead to users taking the underscore less seriously.

infinisil replied:
> just like `meta`

`meta` (and the deprecated `require`) are the only examples of attributes only available in non-shorthand syntax. And `meta` is only that way because [originally](https://github.com/NixOS/nixpkgs/pull/15025) it was `config.meta`.

At least one problem with supporting attributes only in non-shorthand syntax is that it has a poor error messages:

```nix
error: The option `class' does not exist
```

Because of this I'd still prefer `_class`, but then make it work in both syntaxes

roberth replied:
Done.

infinisil commented on lib/modules.nix:
```diff
@@ -256,6 +256,7 @@ rec {
 
       merged =
         let collected = collectModules
+          specialArgs.class or null
```
I don't see a good reason why this should be a `specialArgs` (which gives modules access to that value) instead of just another argument to `evalModules` (and then also `submoduleWith`).

roberth replied:
It is cheap for the general case of having no class; most submodules don't.
Having this in the module arguments allows modules to adapt to the class they're loaded into, which seems useful.

infinisil replied:
I don't see why it would be expensive to have a separate argument for this.

And using `specialArgs` for this seems wrong to me, they're "special" args because they're just like custom arguments, but more infinite-recursion safe. This change would make `.class` have special semantics relevant for the module system evaluation, which breaks that pattern. It could also break user code fwiw.

> which seems useful.

I'd rather we not expose this to modules at first, unless we have a clear need for it. And if we do have a clear need for it, I'd expose it as `config._module.class` instead, because that definitely won't break user code.

roberth replied:
Done.

infinisil reviewed (changes requested):
I took another close look at this and I think I'm onboard with this idea, though I still have some things to point out.

And in addition to the points below, this should also have documentation.

roberth committed:
lib/modules: Only interpret class declaration in non-shorthand mode

This is to avoid stealing keys from submodules. `class` might be
common enough that reinterpreting existing `class` attributes in
configurations as a declaration leads to fairly widespread problems.

roberth committed:
nixos,nixpkgs: Add module classes

This allows modules that declare their class to be checked.
While that's not most user modules, frameworks can take advantage
of this by setting declaring the module class for their users.
That way, the mistake of importing a module into the wrong hierarchy
can be reported more clearly in some cases.

roberth commented:
> And in addition to the points below, this should also have documentation.

It's documented to the current standard for module system documentation, which is barely above zero: code comments only.
I'll make an attempt to improve this, but I'm concerned that it will further delay this PR.

roberth committed:
doc: Add Module System chapter start

bew commented:
There doesn’t seem to have been any bikesheding on the `class` name, so here are a few suggestions (with emoji to react with to vote):
- ❤️ `domain` (my favorite at the moment)
- 👍 `variety`
- 👀 `flavor`
- 🎉 `style`
- (😀 if you want to keep `class`)
- (😕 if you have other suggestions)
- (👎 if you don't think we should vote now)

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+#### `modules` {#module-system-lib-evalModules-param-modules}
+
+A list of modules. These are merged together using various methods to form the final configuration.
```
Would be great if we could refer to those methods more concretely somehow. There surely is section on merging?

roberth replied:
I've removed it, as it contained no information. It should be documented, but that's a rabbit hole for another PR.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+An attribute set of module arguments that can be used in `imports`.
+
+This is in contrast to `config._module.args`, which is only available within the module fixpoint, which does not exist before all imports are resolved.
```
If you say fixpoint you lose the last couple of people who were still following. No suggestion how to improve here, I suppose we just need a better introduction to the module system. @NixOS/documentation-team that would be a great, in fact heroic, first task for a NixOS manual maintainer.

roberth replied:
Good point! Rephrased it in terms of what goes before what; no "fixpoint" needed.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+If the `class` attribute is set in `specialArgs`, the module system will rejected modules with a different `class`.
+
+This improves the error message that users will encounter when they import an incompatible module that was designed for a different class of configurations.
```
```suggestion
```


Sounds like release notes material.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+This improves the error message that users will encounter when they import an incompatible module that was designed for a different class of configurations.
+
+The `class` value should be in camelcase, and, if applicable, it should match the prefix of the attributes used in (experimental) flakes. Some examples are:
```
```suggestion
The `class` value should be in camelcase.

If applicable, it should match the prefix of the attributes used in flakes (experimental). Examples:
```

I'm not sure what you mean by prefix though. Is it the attrset name containing the actual module?

Please link to a sensible definition of camel case. Also link to flakes, otherwise this will be another instance where people will be left wondering what those damn flakes are even supposed to be.

roberth replied:
I've put scare quotes around prefix to hint that it's a local, informal definition. An example follows and explains it.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+#### `options` {#module-system-lib-evalModules-return-value-options}
+
+The nested set of all option declarations.
```
```suggestion
The nested attribute set of all option declarations.
```

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+#### `type` {#module-system-lib-evalModules-return-value-type}
+
+A module system type representing the module set as a submodule, to be extended by configuration from the containing module set.
```
I... don't really understand what that's supposed to mean. Can you provide an example or link to definitions of terms? There seem to be at least three different concepts at play:
1. Types in the module system
2. Submodules
3. Extending one of these

roberth replied:
I've split it into multiple sentences and elaborated it more. Would much appreciate another look.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+A module system type representing the module set as a submodule, to be extended by configuration from the containing module set.
+
+This is also available as the module argument `moduleType`.
```
Is it available *to* a module as an argument? Can you link to docs where that is shown?

roberth replied:
Yes, to a module arguments, but those are currently undocumented, and doing that is too many rabbit holes deep for now.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+#### `extendModules` {#module-system-lib-evalModules-return-value-extendModules}
+
+A function similar to `evalModules` but building on top of the module set. Its arguments, `modules` and `specialArgs` are added to the existing values.
```
Which module set? The returned one?

```suggestion
A function similar to `evalModules` but building on top of this module attribute set. Its arguments, `modules` and `specialArgs` are added to the existing values.
```

roberth replied:
It's funny that we don't have a term for an _actual set_ of modules. I think I'll rely on the plural and a link to the parameter.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+A function similar to `evalModules` but building on top of the module set. Its arguments, `modules` and `specialArgs` are added to the existing values.
+
+Using `extendModules` a few times has no performance impact as long as you only reference the final `options` and `config`.
```
```suggestion
Using `extendModules` multiple times has no impact on evaluation performance as long as you only reference the final `options` and `config`.
```

I know that people not firm with the Nix language will be left puzzled if you mean NixOS runtime performance.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+A function similar to `evalModules` but building on top of the module set. Its arguments, `modules` and `specialArgs` are added to the existing values.
+
+Using `extendModules` a few times has no performance impact as long as you only reference the final `options` and `config`.
+If you do reference multiple `config` (or `options`) from before and after `extendModules`, performance is the same as with multiple `evalModules` invocations, because the new modules' ability to override existing configuration fundamentally requires a new fixpoint to be constructed.
```
```suggestion
If you do reference multiple `config` (or `options`) from before and after `extendModules`, evaluation performance is the same as with multiple `evalModules` invocations, because the new modules' ability to override existing configuration requires a constructing a new fixpoint.
```

This may create more confusion than it alleviates, but let's lebe it here in hope for a good introduction.

roberth replied:
I've put it in a `{.note}` with an **Evaluation Performance** header in it, so that it's clear that this is not required reading.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+Using `extendModules` a few times has no performance impact as long as you only reference the final `options` and `config`.
+If you do reference multiple `config` (or `options`) from before and after `extendModules`, performance is the same as with multiple `evalModules` invocations, because the new modules' ability to override existing configuration fundamentally requires a new fixpoint to be constructed.
+
+This is also available as a module argument.
```
As above, which module argument, where?

roberth replied:
Still out of scope.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+#### `_module` {#module-system-lib-evalModules-return-value-_module}
+
+A portion of the configuration tree which is elided from `config`. It contains some values that are mostly internal to the module system implementation.
```
Should it be documented then? One could argue this shouldn't be exposed to begin with.

roberth replied:
This is tech debt until the markdown migration is complete and we can show the `_module` docs again.
Internal-ness has been debated.

infinisil replied:
I don't think `_module` should be internal. As I can see, the main reason it should be prefixed with a `_` is to avoid clashing with users option names. In https://github.com/NixOS/nixpkgs/pull/97023 I even tried enabled rendering of these options in the manual. That entire PR was reverted, but that as a separate change would still be good.

fricklerhandwerk reviewed (commented):
Docs could be slightly clarified, but overall a great addition of information that's direly missing.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-04-13-documentation-team-meeting-notes-41/27264/1

roberth committed:
lib/modules.nix: Deduplicate documentation

`file://./..` looks redundant, but makes the url clickable in vscode.

roberth commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,69 @@
+
+### Parameters {#module-system-lib-evalModules-parameters}
+
+#### `modules` {#module-system-lib-evalModules-param-modules}
```
Having anchors is more important than the exact look, but in my opinion an `h4` header is already rendered like a definition list; same size font, but not tabular I guess.

roberth committed:
module-system.chapter.md: Add mental model to `type` and `extendModules`

roberth committed:
module-system.chapter.md: Elaborate on extendModules performance

roberth committed:
lib/modules: Move class out of specialArgs

roberth committed:
lib/tests/modules.sh: Add submodule + class tests

roberth committed:
lib.modules: Change class declaration in module to _class

infinisil commented on lib/modules.nix:
```diff
@@ -110,8 +79,12 @@ rec {
                   specialArgs ? {}
-                , # This would be remove in the future, Prefer _module.args option instead.
-                  args ? {}
+                , # `class`:
+                  # A nominal type for modules. When set and non-null, this adds a check to
+                  # make sure that only compatible modules are imported.
+                  # This would be remove in the future, Prefer _module.args option instead.
+                  class ? null
+                , args ? {}
```
```suggestion
                , # A nominal type for modules. When set and non-null, this adds a check to
                  # make sure that only compatible modules are imported.
                  class ? null
                , # This would be remove in the future, Prefer _module.args option instead.
                  args ? {}
```

roberth replied:
Thanks!

infinisil commented:
> It's documented to the current standard for module system documentation, which is barely above zero: code comments only.
> I'll make an attempt to improve this, but I'm concerned that it will further delay this PR.

Yeah that's true. In this case I think we shouldn't bother with documentation in this PR. It's weird to have documentation on the module system but only just enough to a explain a very specific advanced feature that is only just introduced.

infinisil commented on lib/types.nix:
```diff
@@ -791,6 +797,46 @@ rec {
         };
       };
 
+    configuration = { class ? null, allowUnknownClass ? false, allowLegacyType ? false }:
```
This feels out of scope for this PR, it doesn't appear needed for `class` to work, I'd just remove this

roberth replied:
If we merge this simultaneously, we can use `types?configuration` as a simple test whether classes are supported.
I agree that it is technically separable though. Is my reason for inclusion good enough?

infinisil replied:
I don't think so :sweat_smile:. We can already check for classes support now with `functionArgs evalModules ? class` (though maybe we want more clear versioning scheme in the future, maybe `lib.modules.interfaceVersion` or so.).

roberth commented:
> > It's documented to the current standard for module system documentation, which is barely above zero: code comments only.
> > I'll make an attempt to improve this, but I'm concerned that it will further delay this PR.
> 
> Yeah that's true. In this case I think we shouldn't bother with documentation in this PR. It's weird to have documentation on the module system but only just enough to a explain a very specific advanced feature that is only just introduced.

Do you think it actually ended up weird? I think `lib.evalModules` is a reasonable initial scope for the docs. I could un-`include` the file if you think it isn't ready for viewing by users, but to me that feels counterproductive. Something is better than nothing.

infinisil commented:
> Do you think it actually ended up weird?

Hmm yeah somewhat, because the covered concepts are not very relevant to most module system users. But in the docs team we also agreed to really just have reference docs in the NixOS/Nixpkgs/Nix manuals, so it actually fits into that model. Tutorials for new users should be in another place then.

So yes, I take that back, let's have these docs :)

roberth committed:
doc/module-system: `_module` is not internal

nixos-discourse commented:
This issue has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-47/27387/1

roberth commented:
Anything left to do here?

(Except documenting the rest of the module system, which is a project of its own)

infinisil commented:
I'd like to do another review, but I don't always have the time for that. I'll try to get to it within the next week though, ping me if I forget

infinisil commented on lib/modules.nix:
```diff
@@ -110,8 +79,12 @@ rec {
+                  # A nominal type for modules. When set and non-null, this adds a check to
+                  # make sure that only compatible modules are imported.
+                  # This would be remove in the future, Prefer _module.args option instead.
+                  class ? null
```
How about changing this to `expectedClass` (and `_module.expectedClass`), in order to distinguish it from the special `_class` per-module attribute, otherwise these are really easy to confuse.

roberth replied:
Although we've scoped it out for this PR, this won't be quite as applicable when `types.configuration` is added. In that use case `expectedClass` would have the role of the actual class in the check comparison.
Like a type, I see it as more of an intrinsic property than a detail that serves a single check.

infinisil replied:
I'm not sure I get the connection to `types.configuration`, can you elaborate? I also don't yet know why `types.configuration` would be needed.

roberth replied:
`evalModules` takes modules with a `_class` and checks it against `_module.class`.
`types.configuration` takes `_module.class` from the `evalModules` return value and checks it against the type's `class`.

> why `types.configuration` would be needed

Submodule is generally preferable, but `evalModules` results do occur in flake-parts; for example [`nixosConfigurations`](https://flake.parts/options/flake-parts.html#opt-flake.nixosConfigurations).

I wouldn't mind defining the type elsewhere, but `lib.types` is a natural place for it.
Crucially though, the `(evalModules _)._module.class` is an "actual" class, and is not used as an expected class anymore, regardless of `types.configuration`.

infinisil replied:
Hmm I see, that makes sense.

However I think `_module.*` should only be treated as inputs to the module system evaluation, not outputs. It also doesn't make sense for `_module.args` or `_module.check` to be a property of an evaluated configuration, at most to inspect/replicate how evaluation happened.

How about this: Call the argument and option `expectedClass`, but additionally expose it in the result as `(evalModules ...).class`. Or maybe even better: `.configurationClass`, relating more closely to `._type == "configuration"`.

roberth replied:
I've implemented the `(evalModules _).class` idea, but reverted the rename. Having to explain that these `*Class` names all refer to the same thing is not so great. The role of the class variable can be inferred from context, and since they have to be equal anyway, it doesn't even matter much which is which. Adding qualifiers in front of it unnecessarily eats brain space.
The `configuration` qualifier (in `configurationClass`) is implied by the `_type`, so I don't think we need to repeat it. It's unlikely to be ambiguous with the domain, and unlikely to conflict with a potential other framework, which should probably also use `_` prefixes for its framework-specific fields.

infinisil commented on lib/modules.nix:
```diff
@@ -1218,4 +1231,67 @@ rec {
   };
+
+  private = lib.mapAttrs
+    (k: lib.warn "External use of `lib.modules.${k}` is deprecated. If your use case isn't covered by non-deprecated functions, we'd like to know more and perhaps support your use case well, instead of providing access to these low level functions.")
```
```suggestion
    (k: lib.warn "External use of `lib.modules.${k}` is deprecated. If your use case isn't covered by non-deprecated functions, we'd like to know more and perhaps support your use case well, instead of providing access to these low level functions. In this case please open an issue in https://github.com/nixos/nixpkgs/issues/.")
```

infinisil reviewed (changes requested):
I think this looks good other than the points below :)

roberth committed:
lib.modules: in evalModules return move _module.class -> configurationClass

roberth committed:
lib.modules: configurationClass -> class

This simplifies the documentation. `configuration` is implied by `_type`.

infinisil commented on lib/modules.nix:
```diff
@@ -336,38 +310,64 @@ rec {
+            throw (
+              "Could not load a value as a module, because it is of type ${lib.strings.escapeNixString m._type}"
+              + lib.optionalString (fallbackFile != unknownModule) ", in file ${toString fallbackFile}."
+              + lib.optionalString (m._type == "configuration") "If you do intend to import this configuration, please only import the modules that make up the configuration. You may have to create a `let` binding, file or attribute to give yourself access to the relevant modules.\nWhile loading a configuration into the module system is a very sensible idea, it can not be done cleanly in practice ...
```
```suggestion
              + lib.optionalString (m._type == "configuration") " If you do intend to import this configuration, please only import the modules that make up the configuration. You may have to create a `let` binding, file or attribute to give yourself access to the relevant modules.\nWhile loading a configuration into the module system is a very sensible idea, it can not be done cleanly in practice."
```

infinisil commented on lib/types.nix:
```diff
@@ -691,11 +685,18 @@ rec {
         in mergedOption.type;
     };
 
+    # A submodule (like typed attribute set). See NixOS manual.
+    submodule = modules: submoduleWith {
+      shorthandOnlyDefinesConfig = true;
+      modules = toList modules;
+    };
+
```
This code has been moved around unnecessarily

infinisil reviewed (approved):
Other than the minor suggestions, I'd appreciate some squashing for mostly bisection and historical purposes.

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,105 @@
+
+## `lib.evalModules` {#module-system-lib-evalModules}
+
+Evaluate a set of modules. This function is typically only used once per application (e.g. once in NixOS, once in home-manager, ...).
```
```suggestion
Evaluate a set of modules. This function is typically only used once per application (e.g. once in NixOS, once in Home Manager, ...).
```

Apparently this is the proper name

fricklerhandwerk commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,105 @@
+
+#### `class` {#module-system-lib-evalModules-param-class}
+
+If the `class` attribute is set and non-`null`, the module system will reject `imports` with a different `_class` declaration.
```
```suggestion
If the `class` attribute is set and not `null`, the module system will reject `imports` with a different `_class` declaration.
```

Link to _class? It's a different one it seems, so needs a reference.

fricklerhandwerk reviewed (approved):
Docs LGTM modulo nits.

roberth commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,105 @@
+
+An attribute set of module arguments that can be used in `imports`.
+
+This is in contrast to `config._module.args`, which is only available after all `imports` have been resolved.
```
TBD, issue
 - https://github.com/NixOS/nixpkgs/issues/230361

roberth commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,105 @@
+
+#### `prefix` {#module-system-lib-evalModules-param-prefix}
+
+A list of strings representing the location at or below which all options are evaluated. This is used by `types.submodule` to improve error reporting and find the implicit `name` module argument.
```
Not linkable because it's rendered as a stupid variable list without anchors.
I'd change it to be all headers, but this PR has enough scope creep.

roberth commented on doc/module-system/module-system.chapter.md:
```diff
@@ -0,0 +1,105 @@
+
+A function similar to `evalModules` but building on top of the already passed [`modules`](#module-system-lib-evalModules-param-modules). Its arguments, `modules` and `specialArgs` are added to the existing values.
+
+You can think of the current, actual `evalModules` invocation as the prototype, and the return value of `extendModules` as the instance.
+
```
It makes sense in the context of prototype inheritance, so I've added "If you're familiar with prototype inheritance,".

roberth committed:
fixup! doc: Add Module System chapter start

roberth commented:
Thanks for the many reviews :heart:

roberth commented:
Here's a PR for the postponed `types.configuration`
-  https://github.com/NixOS/nixpkgs/pull/230377

uninsane commented:
> External use of `lib.modules.pushDownProperties` is deprecated. If your use case isn't covered by non-deprecated functions, we'd like to know more and perhaps support your use case well, instead of providing access to these low level functions.

@roberth as far as i can tell, my use case isn't covered by non-deprecated functions. as you expected, external uses of these are a bit gnarly, but i'll try here to condense my justification for its use.

`config = mkMerge <list>` at module scope often causes infinite recursion, e.g. when the items in `<list>` are computed attrsets: a well-known issue but with AFAIK no widely agreed upon solution. code like this barfs:

```nix
{ config, ... }:
{
  options.my.option = mkOption {
    type = types.listOf types.string;
  };
  
  config = mkMerge (map generateMoreConfig config.my.option);
  # ^ ERROR: `config.*` depends on `config.my.option` => infinite recursion
}
```

assume `generateMoreConfig` produces an attrset with `assertions` and `users`. the simple fix is:
```nix
  config = let
    genAssertions = conf: (generateMoreConfig conf).assertions;
    genUsers = conf: (generateMoreConfig conf).users;
  in {
    assertions = mkMerge (map genAssertions config.my.option);
    users = mkMerge (map genUsers config.my.option);
  };
```

however, this **fails** if `generateMoreConfig` includes any `lib.mkIf` at the toplevel, e.g.
```nix
generateMoreConfig = value: lib.mkIf value != "root" {
  assertions = ...; users = ...;
};
```

hence, any more robust equivalent to module-scope `mkMerge` has to "push down" that outer `lib.mkIf`:
```nix
generateMoreConfig = value: {
  assertions = lib.mkIf value != "root" ...;
  users = lib.mkIf value != "root" ...;
};
```

but you don't want to manually push these down. ideally one could take the original, clean-but-unsupported code:
```nix
config = mkMerge (map generateMoreConfig config.my.option);
```

and defeat the infinite recursion by constraining the attrnames at the outer layer -- _without_ contorting the logic everywhere below that:
```nix
config = let
  mergeEachAttr (map generateMoreConfig config.my.option);
in {
  assertions = mergeEachAttr.assertions;
  users = mergeEachAttr.users;
};
```

hence, this `mergeEachAttr` function has to use `lib.pushDownProperties`. in case anyone wants to dive deeper, my concrete use:
- helper definition: [mkTypedMerge](https://git.uninsane.org/colin/nix-files/src/commit/ada8b756707f1a72441a9d7afefdc6bd38bb55d2/modules/lib/merge.nix#L13)
- [where it's used](https://git.uninsane.org/colin/nix-files/src/commit/ada8b756707f1a72441a9d7afefdc6bd38bb55d2/modules/fs/default.nix#L358)

roberth commented:
@uninsane That seems like a valid use case. Could you create an issue for it?

uninsane commented:
> @uninsane That seems like a valid use case. Could you create an issue for it?

the `mkMerge` limitations are i think well-known (but with no foreseeable direct "fix"). unless i'm not thinking in the right scope, the only new/actionable issue i can see is "`pushDownProperties` was marked private even though a single external user still depends on it". but i don't know that that's even an _issue_: i only felt compelled to say anything about it because you signaled in the code change that you'd like to hear from anyone who's using these. i assumed the code committer would then be in a position to make any relevant judgements.

if you prefer i formalize that feedback through some issue, i'm happy to do that, but i'd need guidance because i have so much less context here than you that i don't know what the substance of any issue here actually _is_.

roberth commented:
My goal is just to have a separate thread for it, not to put you through the bureaucratic wringer or something. This thread already has over a hundred messages, and changing the topic of a long thread makes things harder to find. GitHub hiding the middle also doesn't help.

> limitations are i think well-known (but with no foreseeable direct "fix")

While there's an element of there being a fundamental problem that prevents the simplest fix imaginable, that doesn't stop is from finding practical solutions, like you did!

> but i'd need guidance because i have so much less context

Most issues are conversations. They don't have to be full-on research presentations. You might fill in the template and paste your existing prose in the appropriate header(s). Markdown source is under the comment edit button ;)
